### DISCUSSION TITLE
[Question] How to switch .env for migration

### DISCUSSION THREAD

**Discussion by hyochan** [prisma/prisma#2392] [Q&A] [answered]
```
// schema.prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}
```

How can I switch `DATABASE_URL` in `schema.prisma` in runtime? Say I want to use a test database and dev database or production database for migration.

**Comment by ryands17** [comment, marked as answer]
Hey @hyochan :wave: 

Instead of switching the variable `DATABASE_URL`, it would be better if you switch the `.env` file itself using [dotenv-cli](https://www.npmjs.com/package/dotenv-cli)

For e.g.

```
dotenv -e .env.prod -- npx prisma migrate up --experimental
```

To add the migrations to the production DB using a different environment.

**Reply by hyochan** [reply]
@ryands17 Looks like this is currently a nice workaround! Thanks for sharing~

**Comment by smallbee3** [comment]
It seems that the dynamic datasource feature is not available now.
However, during searching for issues with the same topic, I found out that it would be developed soon as Beta. #1487 

But as ryands17 mentioned above, dotenv-cli looks a possible option now!

**Reply by hyochan** [reply]
Looks like you've found the related issue! Thanks

**Comment by hyochan** [comment]
@ryands17 I've got another problem using this solution. According to the [boilerplate](https://github.com/dooboolab/ts-apollo-prisma) I am building today, I currently have no idea how to separate the test files. The [testSetup](https://github.com/dooboolab/ts-apollo-prisma/blob/master/tests/testSetup.ts) runs `beforeAll` and `afterAll` methods to run the server and in `afterAll`, it drops the schema as you've recommended.

However, we need to run `prisma migrate up` in `beforeAll` with `test.env` and it doesn't seem to make this possible currently. How would you prefer in this case?

P.S. Sorry if my explanation is not clear enough. Please tell me then I'll try again.

**Reply by ryands17** [reply]
@hyochan I can see the `test` script in the repo. Is that what you're running currently?
And are you facing any issues with `prisma migrate up` while doing that?

**Reply by hyochan** [reply]
I am not facing any issue with `prisma migrate up`. My remaining issue is that I can't run `dotenv` cli inside `beforeAll` in `testSetup.ts`.

If the above is possible, I can separate my one big test file into many.

**Reply by ryands17** [reply]
What would be the use-case for running that inside `beforeAll`, would it be to access environment variables or for something else. Could you provide an example if possible?

**Reply by hyochan** [reply]
![image](https://user-images.githubusercontent.com/27461460/82662108-3fd9ed80-9c68-11ea-8f12-5baf850e719c.png)

Since there wasn't any real example to make the test `env` for `prisma`, I'd considered writing like above in `testSetup.ts`. At least I can test e2e resolvers.

**Reply by ryands17** [reply]
So when you append `dotenv-cli` to the test script, aren't the env variables available in the `beforeAll` block?

**Reply by hyochan** [reply]
![image](https://user-images.githubusercontent.com/27461460/82676944-2c3b8080-9c82-11ea-9139-3eba4a279a9b.png)

If we add more tests like above with multiple files, it won't work since in `afterAll`, the prisma will drop the `schema`. We need to re-run the `dotenv-cli` with `test.env` and run `primsa migrate` so that other tests can proceed.

**Reply by ryands17** [reply]
I think the best approach here would be running `prisma migrate` in `beforeAll`.
Using something like `exec`.

```ts
beforeAll((done) => {
    exec(`npm run test:migrate`, (err, stdout) => {
      console.log({ err, stdout });
     done();
    });
});
```

Where the script `migrate` could be:
```json
"test:migrate": "dotenv -e test.env -- prisma migrate up --experimental"
```

**Reply by dan-kez** [reply]
As a followup, you can run into conflicts here if you wish to run your tests in parallel.

It would be nice to be able to dynamically change the database name in the `beforeAll` script so that each file could get it's own db and not compete.

Additionally, It would be great if some common recipes for end to end tests were included in the documentation. It's a fairly common use case to want to write tests against service methods that have an effect on the db and assert that changes have taken affect.

**Reply by hyochan** [reply]
@dan-kez That's such a good consideration. In addition, the approach to setting up the database url source as below seems pretty ridiculous.
```
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}
```

We should able to handle above instead of `dotenv-cli`.

**Reply by ryands17** [reply]
@hyochan Was it possible for you to perform the `beforeAll` workaround?
If so, is that mentioned in a repository someplace?

**Reply by hyochan** [reply]
@ryands17 Sorry that I've missed your reply. [Here](https://github.com/dooboolab/hackatalk/blob/fe914e576b00ffd23a46b9deb06e34ae8bc62e5b/server/tests/setup/testSetup.ts#L35-L38) is our repo! It is an opensource.

**Comment by dsebastien** [comment]
The dotenv-cli workaround works, but the Prisma CLI incorrectly states the following: `Environment variables loaded from .env`, even though it DID load env variables from another file. Not a biggie though :)

**Comment by ludralph** [comment]
Hi @hyochan,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
